### PR TITLE
Tickets/pipe2 d 537

### PR DESCRIPTION
**Pull request by takujiyamashita**
The template interpolation of stellar models is performed with an RBF model. Because the codes for making and reading the RBF model are implemented, the codes and a related document should be reviewed. These are the first version based on the current study. The files are as follows.

./tools/makeRBFInterpolationModel.py
./tools/modelInterpolation.py
./doc/memo_interpolation.mdwn

### CODE REVIEW

File: doc/memo_interpolation.mdwn [hunk 1]
```diff
@@ -1,26 +1,25 @@
...
 <!-- ### About AMBRE -->
-AMBRE is the high resolution stellar model templates (de Laverny et al. 2012, A&A, 544, 126). It covers a wavelength range from 300 nm to 1200 nm. The original resolution is 0.001 nm/pix. There are four parameters of Teff, log(g), Z, [alphe/Fe]. 
+AMBRE is a high resolution stellar model templates (de Laverny et al. 2012, A&A, 544, 126). The wavelength coverage is from 300 nm to 1200 nm. The original resolution is 0.001 nm/pix. There are four parameters of Teff, log(g), Z, [alpha/Fe]. 
 
 <!-- ### Input data for interpolation -->
 Out of the AMBRE templates, 3190 model templates with Teff = 5500 - 8000 K and plane-parallel atmosphere are used as input data of the interpolation.
-The wavelength pixel of the input templates are binned into 0.01 nm/pix to reduce a number of calculation.
-They are stored as a FITS format file of 0.4 MB size. In total, 1.27 GB.
-
+The spectral resolution is reduced down to 0.01 nm/pix by pixel binning to reduce a number of calculation.
```
Comment (hassanxp) [inline doc/memo_interpolation.mdwn]: [...] to reduce the number of calculations [...]

File: tools/ambreModelInterpolation4DRBF.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+
+class AmbreInterpolation: 
+    """ Interpolate a model spectrum along four model parameters using scipy.interpolate.Rbf.
+        `interpolate` method returns a `numpy.ndarray` of an interpolated spectrum.
+        This script calls `parallel.py`.
+
+    Parameters
+    ----------
+    teff : Effective temperature (Teff [K]) at an interpolation. type = float
```
Comment (hassanxp) [inline tools/ambreModelInterpolation4DRBF.py]: Should be formatted as:
``
teff : `float`
    Effective temperature (T_eff [K]) at an interpolation point
``

And the same for the other parameters.
See https://developer.lsst.io/python/numpydoc.html#py-docstring-basics for more info/examples

File: tools/ambreModelInterpolation4DRBF.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                                           (parameterList['Alpha'] == self.alpha)]['ModelName'].values
+        if referenceModelName.size != 0:
+            return referenceModelName[0]
+        if referenceModelName.size == 0:
+            return np.array([])
+
+    def interpolate(self, nProcs=1):        
+        """ Do interpolation
+
+        Parameter
```
Comment (hassanxp) [inline tools/ambreModelInterpolation4DRBF.py]: Even though there is a single parameter here, the docString section should be 'Parameters'.

File: tools/ambreModelInterpolation4DRBF.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        ---------
+        --nProcs : Number of processes for multiprocessing; the default is 1 (sigle process). Type = int
+        """
+        def fetchNeighborModels(self, indexTeffRange, indexLoggRange, indexMetalRange, indexAlphaRange, 
+                                referenceModelName):
+            """ Fetch a dataset of model names and parameters of neighbor models.
+            If there is a reference model, it is removed from the neighbor model list.
```
Comment (hassanxp) [inline tools/ambreModelInterpolation4DRBF.py]: Need to document all parameters and return value.

File: tools/ambreModelInterpolation4DRBF.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            (numOfModelsAlongAlpha >= requiredNumberOfParameters)): 
+            neighborModel = fetchNeighborModels(self, indexTeffRange, indexLoggRange, indexMetalRange, 
+                                                indexAlphaRange, referenceModelName)
+            
+            neighborModelParameter = np.array(neighborModel[['Teff','Gravity', 'Metal', 'Alpha']])
+            neighborModelName = neighborModel['ModelName']
+            
+            with fits.open(self.path2Spectrum + neighborModelName.reset_index(drop=True)[0]) as tempModelFITS:
+                tempModel = Table(tempModelFITS[1].data).to_pandas()
+#            neighborModelFlux = np.empty((len(tempModel), len(neighborModelName)), dtype='float64')
```
Comment (hassanxp) [inline tools/ambreModelInterpolation4DRBF.py]: Remove commented out lines

File: tools/ambreModelInterpolation4DRBF.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            neighborModelFlux = np.empty((100, len(neighborModelName)), dtype='float64')
+            
+            for i, row in enumerate(neighborModelName):
+                with fits.open(self.path2Spectrum + row) as modelFITS:
+                    model = Table(modelFITS[1].data).to_pandas()
+#                neighborModelFlux[:,i] = model['Flux']
```
Comment (hassanxp) [inline tools/ambreModelInterpolation4DRBF.py]: Remove this commented out line also

File: tools/ambreModelInterpolation4DRBF.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+                                            neighborModelFlux, function=self.function, epsilon=self.epsilon)
+                interpolatedFlux = interpolationFunction(self.teff/1e3, self.logg, self.metal, self.alpha)
+                return interpolatedFlux
+
+            result = parallel.parallel_map(rbfInterpolate, neighborModelFlux, n_procs=nProcs)
+            result = np.hstack(result)           
+            return result
+
+        else:
+            print('Not enough number of parameters. This process ends')
```
Comment (hassanxp) [inline tools/ambreModelInterpolation4DRBF.py]: Can you replace ``print`` calls with ``lsst.logger`` calls? https://developer.lsst.io/stack/logging.html

File: tools/ambreModelInterpolation2DRBF.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        return len(Ambre2drbfInterpolation.originalParam(self))
+
+    def interpolate(self, nProcs=1):
+        """ Do interpolation
+        Return an interpolated spectrum
+        Usage: interpolation = interpolate(nProcs)
+        Parameter: 
+        --nProcs : Number of processes for multiprocessing; the default is 1 (sigle process). Type = int
+        """
+
```
Comment (hassanxp) [inline tools/ambreModelInterpolation2DRBF.py:47]: Remove this blank line

File: tools/ambreModelInterpolation2DRBF.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    def originalParamNum(self):
+        """ Count a number of original grid points """
+        return len(Ambre2drbfInterpolation.originalParam(self))
+
+    def interpolate(self, nProcs=1):
+        """ Do interpolation
+        Return an interpolated spectrum
+        Usage: interpolation = interpolate(nProcs)
+        Parameter: 
+        --nProcs : Number of processes for multiprocessing; the default is 1 (sigle process). Type = int
```
Comment (hassanxp) [inline tools/ambreModelInterpolation2DRBF.py:45]: Typo: ``sigle`` -> ``single``. Also remove the ``--`` prefix to the parameter name and format according to the LSST docString standards.

File: tools/modelInterpolation.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import numpy as np
+import parallel
+import pickle
+
+class InterpolationModel:
+    def __init__(self, modelPickle):
```
Comment (hassanxp) [inline tools/modelInterpolation.py:19]: This class (and second class in this file) need docStrings to describe their purpose.

File: doc/memo_interpolation.mdwn (hunk 1 above)
Comment (takujiyamashita) [reply doc/memo_interpolation.mdwn]: done.

File: tools/ambreModelInterpolation4DRBF.py (hunk 2 above)
Comment (takujiyamashita) [reply tools/ambreModelInterpolation4DRBF.py]: Different code was reviewed.

File: tools/makeRBFInterpolationModel.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+import parallel
+import pickle
+
+
+""" Set input information """
+## inputDir: Directory name of input spectra.
+inputDir = '/data22b/work/takuji/AMBRE/extrapolation20190419/extrapolated_all/'
+## inputList: Ascii table listing parameters to be used for interpolation.
+inputList = '/home/takuji/work/ambre/jn/extrapolatedmodel_parameters_p5500-8000.dat'
+## output: Output file name.
```
Comment (hassanxp) [inline tools/makeRBFInterpolationModel.py]: These (and in particular ``inputDir`` and ``inputList`` are local configuration parameters hard-coded. Could the code be updated to accept such parameters a command-line arguments and pass the hard-coded values from a shell script?

File: tools/makeRBFInterpolationModel.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+inputDir = '/data22b/work/takuji/AMBRE/extrapolation20190419/extrapolated_all/'
+## inputList: Ascii table listing parameters to be used for interpolation.
+inputList = '/home/takuji/work/ambre/jn/extrapolatedmodel_parameters_p5500-8000.dat'
+## output: Output file name. 
+output = 'ambreRBF.pickle'
+## n_procs: int. Number of processes. 
+n_procs = 7
+
+
+def makeRBFInterpolationModel(inputDir, inputList, output, n_procs):
```
Comment (hassanxp) [inline tools/makeRBFInterpolationModel.py]: Could you provide a docString that describes the purpose of this function and the input parameters? See https://developer.lsst.io/python/numpydoc.html for information.

**Review by hassanxp — COMMENTED** [review]
Can you look at my latest comments and update please?

File: tools/modelInterpolation.py (hunk 10 above)
Comment (takujiyamashita) [reply tools/modelInterpolation.py:19]: Done.

File: tools/makeRBFInterpolationModel.py (hunk 11 above)
Comment (takujiyamashita) [reply tools/makeRBFInterpolationModel.py]: Done.

File: tools/makeRBFInterpolationModel.py (hunk 12 above)
Comment (takujiyamashita) [reply tools/makeRBFInterpolationModel.py]: Done.

File: tools/ambreModelInterpolation2DRBF.py (hunk 9 above)
Comment (hassanxp) [reply tools/ambreModelInterpolation2DRBF.py:45]: Has this comment been implemented?

File: tools/ambreModelInterpolation2DRBF.py (hunk 8 above)
Comment (hassanxp) [reply tools/ambreModelInterpolation2DRBF.py:47]: Can this line be removed?

**Review by hassanxp — APPROVED** [review]
Minor outstanding style comments, but otherwise ok.

**Pull request closed without merging**